ultravox/data/datasets.py

```diff
@@ -1138,6 +1141,9 @@ def __iter__(self):
                 iters[iter_index] = iter(self._datasets[iter_index])
                 yield next(iters[iter_index])
 
+    def __len__(self):
+        return sum(len(ds) for ds in self._datasets)
+
```

Comment:
The definition of epoch for the interleaved dataset is a bit tricky due to the weighting of datasets and the stop strategy. I guess the logic used here is likely the most accurate, especially if we consider concatenation as a special case of an interleaved dataset.

Currently, concatenation can be simulated by manually setting the weights of each dataset based on its length. Perhaps we could consider using weight=0 as the default, and if any weight is set to 0, it should trigger concatenation simulation. This could be addressed in a separate PR.

Reply:
I think this depends on how huggingface uses len(train_dataset)? I did some digging and found one place where they use the length of the dataset to determine the number of shards to batch in training. Seems like that would mean we'd need the actual* size of the dataset, including the logic in interleave, and concatenation is not ideal here for strategys like last_exauhsted because it would underestimate the length.